File: platform/platform-api/src/com/intellij/ui/AutoScrollToSourceHandler.java
// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package com.intellij.ui;

import com.intellij.icons.AllIcons;
import com.intellij.ide.DataManager;
import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.ToggleAction;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.openapi.fileTypes.FileTypes;
import com.intellij.openapi.fileTypes.INativeFileType;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.util.NlsActions;
import com.intellij.openapi.util.registry.Registry;
import com.intellij.openapi.vfs.PersistentFSConstants;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.Alarm;
import com.intellij.util.concurrency.annotations.RequiresEdt;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import javax.swing.event.TreeSelectionEvent;
import javax.swing.event.TreeSelectionListener;
import javax.swing.tree.TreePath;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;

public abstract class AutoScrollToSourceHandler {
  private final Alarm myAutoScrollAlarm = new Alarm();

  public void install(final JTree tree) {
    new ClickListener() {
      @Override
      public boolean onClick(@NotNull MouseEvent e, int clickCount) {
        if (clickCount > 1) return false;

        TreePath location = tree.getPathForLocation(e.getPoint().x, e.getPoint().y);
        if (location != null) {
          onMouseClicked(tree);
          // return isAutoScrollMode(); // do not consume event to allow processing by a tree
        }

        return false;
      }
    }.installOn(tree);

    tree.addMouseMotionListener(new MouseMotionAdapter() {
      @Override
      public void mouseDragged(final MouseEvent e) {
        onSelectionChanged(tree);
      }
    });
    tree.addTreeSelectionListener(
      new TreeSelectionListener() {
        @Override
        public void valueChanged(TreeSelectionEvent e) {
          onSelectionChanged(tree);
        }
      }
    );
  }

  public void install(final JTable table) {
    new ClickListener() {
      @Override
      public boolean onClick(@NotNull MouseEvent e, int clickCount) {
        if (clickCount >= 2) return false;

        Component location = table.getComponentAt(e.getPoint());
        if (location != null) {
          onMouseClicked(table);
          return isAutoScrollMode();
        }
        return false;
      }
    }.installOn(table);

    table.addMouseMotionListener(new MouseMotionAdapter() {
      @Override
      public void mouseDragged(final MouseEvent e) {
        onSelectionChanged(table);
      }
    });
    table.getSelectionModel().addListSelectionListener(
      new ListSelectionListener() {
        @Override
        public void valueChanged(ListSelectionEvent e) {
          onSelectionChanged(table);
        }
      }
    );
  }

  public void install(final JList jList) {
    new ClickListener() {
      @Override
      public boolean onClick(@NotNull MouseEvent e, int clickCount) {
        if (clickCount >= 2) return false;
        final Object source = e.getSource();
        final int index = jList.locationToIndex(SwingUtilities.convertPoint(source instanceof Component ? (Component)source : null, e.getPoint(), jList));
        if (index >= 0 && index < jList.getModel().getSize()) {
          onMouseClicked(jList);
          return true;
        }
        return false;
      }
    }.installOn(jList);

    jList.addListSelectionListener(new ListSelectionListener() {
      @Override
      public void valueChanged(ListSelectionEvent e) {
        onSelectionChanged(jList);
      }
    });
  }

  public void cancelAllRequests(){
    myAutoScrollAlarm.cancelAllRequests();
  }

  public void onMouseClicked(final Component component) {
    cancelAllRequests();
    if (isAutoScrollMode()){
      ApplicationManager.getApplication().invokeLater(() -> scrollToSource(component));
    }
  }

  private void onSelectionChanged(final Component component) {
    if (component != null && component.isShowing() && isAutoScrollMode()) {
      myAutoScrollAlarm.cancelAllRequests();
      myAutoScrollAlarm.addRequest(
        () -> {
          if (component.isShowing()) { //for tests
            if (!needToCheckFocus() || component.hasFocus()) {
              scrollToSource(component);
            }
          }
        },
        Registry.intValue("ide.autoscroll.to.source.delay", 100)
      );
    }
  }

  protected @NlsActions.ActionText String getActionName() {
    return UIBundle.message("autoscroll.to.source.action.name");
  }

  protected @NlsActions.ActionDescription String getActionDescription() {
    return UIBundle.message("autoscroll.to.source.action.description");
  }

  protected boolean needToCheckFocus(){
    return true;
  }

  protected abstract boolean isAutoScrollMode();
  protected abstract void setAutoScrollMode(boolean state);

  /**
   * @param file a file selected in a tree
   * @return {@code false} if navigation to the file is prohibited
   */
  protected boolean isAutoScrollEnabledFor(@NotNull VirtualFile file) {
    // Attempt to navigate to the virtual file with unknown file type will show a modal dialog
    // asking to register some file type for this file. This behaviour is undesirable when auto scrolling.
    FileType type = file.getFileType();
    if (type == FileTypes.UNKNOWN || type instanceof INativeFileType) return false;
    //IDEA-84881 Don't autoscroll to very large files
    return file.getLength() <= PersistentFSConstants.getMaxIntellisenseFileSize();
  }

  @RequiresEdt
  protected void scrollToSource(@NotNull Component tree) {
    AutoScrollToSourceTaskManager.getInstance()
      .scheduleScrollToSource(this,
                              DataManager.getInstance().getDataContext(tree));
  }

  @NotNull
  public ToggleAction createToggleAction() {
    return new AutoscrollToSourceAction(getActionName(), getActionDescription());
  }

  private class AutoscrollToSourceAction extends ToggleAction implements DumbAware {
    AutoscrollToSourceAction(@NlsActions.ActionText String actionName, @NlsActions.ActionDescription String actionDescription) {
      super(actionName, actionDescription, AllIcons.General.AutoscrollToSource);
    }

    @Override
    public boolean isSelected(@NotNull AnActionEvent event) {
      return isAutoScrollMode();
    }

    @Override
    public @NotNull ActionUpdateThread getActionUpdateThread() {
      return ActionUpdateThread.EDT;
    }

    @Override
    public void setSelected(@NotNull AnActionEvent event, boolean flag) {
      setAutoScrollMode(flag);
    }
  }
}

File: platform/platform-api/src/com/intellij/ui/AutoScrollToSourceHandler.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.ui

import com.intellij.icons.AllIcons
import com.intellij.ide.DataManager
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.ToggleAction
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.fileTypes.FileTypes
import com.intellij.openapi.fileTypes.INativeFileType
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.util.NlsActions
import com.intellij.openapi.util.registry.Registry
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.isTooLargeForIntellijSense
import com.intellij.util.Alarm
import com.intellij.util.SingleAlarm
import com.intellij.util.concurrency.annotations.RequiresEdt
import com.intellij.util.ui.UIUtil
import org.jetbrains.annotations.ApiStatus
import java.awt.Component
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionAdapter
import java.lang.ref.WeakReference
import javax.swing.JList
import javax.swing.JTable
import javax.swing.JTree
import javax.swing.SwingUtilities

abstract class AutoScrollToSourceHandler {
  private var scheduledNavigationData: WeakReference<Component>? = null

  // access only from EDT
  private var autoScrollAlarm: SingleAlarm? = null 
    
  private fun getOrCreateAutoScrollAlarm(): SingleAlarm {
    var alarm = autoScrollAlarm
    if (alarm == null) {
      alarm = SingleAlarm(
        task = {
          val component = scheduledNavigationData?.get() ?: return@SingleAlarm
          scheduledNavigationData = null
          // for tests
          if (component.isShowing && (!needToCheckFocus() || UIUtil.hasFocus(component))) {
            scrollToSource(component)
          }
        },
        delay = Registry.intValue("ide.autoscroll.to.source.delay", 100),
        parentDisposable = null,
        threadToUse = Alarm.ThreadToUse.SWING_THREAD,
        modalityState = ModalityState.defaultModalityState(),
      )
      autoScrollAlarm = alarm
    }
    return alarm
  }

  fun install(tree: JTree) {
    object : ClickListener() {
      override fun onClick(e: MouseEvent, clickCount: Int): Boolean {
        if (clickCount > 1) {
          return false
        }

        val location = tree.getPathForLocation(e.point.x, e.point.y)
        if (location != null) {
          onMouseClicked(tree)
          // return isAutoScrollMode(); // do not consume event to allow processing by a tree
        }

        return false
      }
    }.installOn(tree)

    tree.addMouseMotionListener(object : MouseMotionAdapter() {
      override fun mouseDragged(e: MouseEvent) {
        onSelectionChanged(tree)
      }
    })
    tree.addTreeSelectionListener { onSelectionChanged(tree) }
  }

  fun install(table: JTable) {
    object : ClickListener() {
      override fun onClick(e: MouseEvent, clickCount: Int): Boolean {
        if (clickCount >= 2) return false

        val location = table.getComponentAt(e.point)
        if (location != null) {
          onMouseClicked(table)
          return isAutoScrollMode()
        }
        return false
      }
    }.installOn(table)

    table.addMouseMotionListener(object : MouseMotionAdapter() {
      override fun mouseDragged(e: MouseEvent) {
        onSelectionChanged(table)
      }
    })
    table.selectionModel.addListSelectionListener { onSelectionChanged(table) }
  }

  fun install(jList: JList<*>) {
    object : ClickListener() {
      override fun onClick(e: MouseEvent, clickCount: Int): Boolean {
        if (clickCount >= 2) return false
        val source = e.source
        val index = jList.locationToIndex(SwingUtilities.convertPoint(if (source is Component) source else null, e.point, jList))
        if (index >= 0 && index < jList.model.size) {
          onMouseClicked(jList)
          return true
        }
        return false
      }
    }.installOn(jList)

    jList.addListSelectionListener { onSelectionChanged(jList) }
  }

  fun cancelAllRequests() {
    autoScrollAlarm?.cancel()
  }

  /**
   * Resets the internal instance to ensure that it'll be recreated properly if the component is reused in a different modal dialog.
   * This is a temporary solution specifically for the Project Structure dialog, it's better to recreate components instead of reusing them.
   */
  @ApiStatus.Internal
  @ApiStatus.Obsolete
  fun resetAlarm() {
    autoScrollAlarm = null
  }

  fun onMouseClicked(component: Component) {
    cancelAllRequests()
    if (isAutoScrollMode()) {
      scrollToSource(component)
    }
  }

  private fun onSelectionChanged(component: Component?) {
    if (component != null && component.isShowing && isAutoScrollMode()) {
      scheduledNavigationData = WeakReference(component)
      getOrCreateAutoScrollAlarm().cancelAndRequest()
    }
  }

  protected open val actionName: @NlsActions.ActionText String?
    get() = UIBundle.message("autoscroll.to.source.action.name")

  protected open val actionDescription: @NlsActions.ActionDescription String?
    get() = UIBundle.message("autoscroll.to.source.action.description")

  protected open fun needToCheckFocus(): Boolean = true

  protected abstract fun isAutoScrollMode(): Boolean

  protected abstract fun setAutoScrollMode(state: Boolean)

  /**
   * @param file a file selected in a tree
   * @return `false` if navigation to the file is prohibited
   */
  @ApiStatus.Internal
  open fun isAutoScrollEnabledFor(file: VirtualFile): Boolean {
    // Attempt to navigate to the virtual file with an unknown file type will show a modal dialog
    // asking to register some file type for this file.
    // This behavior is undesirable when auto scrolling.
    val type = file.fileType
    if (type === FileTypes.UNKNOWN || type is INativeFileType) {
      return false
    }

    //IDEA-84881 Don't autoscroll to very large files
    return !file.isTooLargeForIntellijSense()
  }

  @RequiresEdt
  protected open fun scrollToSource(tree: Component) {
    AutoScrollToSourceTaskManager.getInstance()
      .scheduleScrollToSource(handler = this, dataContext = DataManager.getInstance().getDataContext(tree))
  }

  fun createToggleAction(): ToggleAction = AutoscrollToSourceAction(actionName, actionDescription)

  private inner class AutoscrollToSourceAction(
    actionName: @NlsActions.ActionText String?,
    actionDescription: @NlsActions.ActionDescription String?
  ) : ToggleAction(actionName, actionDescription, AllIcons.General.AutoscrollToSource), DumbAware {
    override fun isSelected(event: AnActionEvent): Boolean = isAutoScrollMode()

    override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT

    override fun setSelected(event: AnActionEvent, flag: Boolean) {
      setAutoScrollMode(flag)
    }
  }
}

